src/type/schema.js

```diff
@@ -147,3 +157,61 @@ function typeMapReducer(map: TypeMap, type: ?GraphQLType): TypeMap {
 
   return reducedMap;
 }
+
+function assertObjectImplementsInterface(
+  object: GraphQLObjectType,
+  iface: GraphQLInterfaceType
+): void {
+  var objectFieldMap = object.getFields();
+  var ifaceFieldMap = iface.getFields();
+  Object.keys(ifaceFieldMap).forEach(fieldName => {
+    var objectField = objectFieldMap[fieldName];
+    var ifaceField = ifaceFieldMap[fieldName];
+
+    invariant(
+      objectField,
+      `"${iface}" expects field "${fieldName}" but "${object}" does not ` +
+      `provide it.`
+    );
+
+    invariant(
+      isEqualType(ifaceField.type, objectField.type),
+      `${iface}.${fieldName} expects type "${ifaceField.type}" but ` +
+      `${object}.${fieldName} provides type "${objectField.type}".`
+    );
+
+    ifaceField.args.forEach(ifaceArg => {
+      var argName = ifaceArg.name;
+      var objectArg = find(objectField.args, arg => arg.name === argName);
+
+      invariant(
+        objectArg,
+        `${iface}.${fieldName} expects argument "${argName}" but ` +
+        `${object}.${fieldName} does not provide it.`
+      );
+
+      invariant(
+        isEqualType(ifaceArg.type, objectArg.type),
+        `${iface}.${fieldName}(${argName}:) expects type "${ifaceArg.type}" ` +
+        `but ${object}.${fieldName}(${argName}:) provides ` +
+        `type "${objectArg.type}".`
+      );
+    });
+  });
+}
+
+function isEqualType(typeA: GraphQLType, typeB: GraphQLType): boolean {
+  if (typeA instanceof GraphQLNonNull) {
+    if (typeB instanceof GraphQLNonNull) {
+      return isEqualType(typeA.ofType, typeB.ofType);
+    }
+    return false;
```

Comment:
Hm, this feels a bit tricky to me, since we return false here when A is NonNull and B is not, but we don't return false here when B is NonNull and A is not.

Perhaps:

```js
if (typeA instanceof GraphQLNonNull && typeB instanceof GraphQLNonNull) {
  return isEqualType(typeA.ofType, typeB.ofType);
}
if (typeA instanceof GraphQLList && typeB instanceof GraphQLList) {
  return isEqualType(typeA.ofType, typeB.ofType);
}
return typeA === typeB;
```

would be clearer?